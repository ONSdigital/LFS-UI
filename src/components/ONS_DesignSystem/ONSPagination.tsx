import React, {Component} from 'react';

interface Props {
    steps: number,
    count: number,
    pageChange: (offset: number, steps: number) => void
}

interface State {
    page: number,
    maxPage: number
}

export class ONSPagination extends Component <Props, State> {
    constructor(props: Props) {
        super(props);
        let maxPage = Math.ceil(props.count / props.steps);
        this.state = {page: 1, maxPage: maxPage}
    }

    static getDerivedStateFromProps(nextProps: Props, prevState: State) {
        let maxPage = Math.ceil(nextProps.count / nextProps.steps);
        if (maxPage !== prevState.maxPage) {
            return {page: 1, maxPage: maxPage};
        } else return null;
    }


    createLinks = () => {

        let prev = [];
        let prevPages = [];
        for (let i = this.state.page - 1; i > 1 && i >= this.state.page - 2; i--) {
            prev.push(this.link(i));
            prevPages.push(i);
        }
        prev = prev.reverse();
        prevPages = prevPages.reverse();
        if (this.state.page !== 1 && this.state.page !== this.state.maxPage) {
            prev.push(this.link(this.state.page));
            prevPages.push(this.state.page)
        }
        let aft = [];
        let aftPages = [];
        for (let i = this.state.page + 1; i < this.state.maxPage && i <= this.state.page + 2; i++) {
            aft.push(this.link(i));
            aftPages.push(i);
        }
        let allPages = prevPages.concat(aftPages);
        let toReturn = prev.concat(aft);

        if (!allPages.includes(2)) {
            toReturn.unshift(<li key={"start_dots"} className="pagination__item pageination__item--gap">&hellip;</li>);
        }
        if (!allPages.includes(this.state.maxPage - 1)) {
            toReturn.push(<li key={"end_dots"} className="pagination__item pageination__item--gap">&hellip;</li>);
        }
        return toReturn
    };

    pageChange = (e: React.MouseEvent, page: number) => {
        if (page !== this.state.page) {
            this.props.pageChange((page - 1) * this.props.steps, this.props.steps);
            this.setState({page: page})
        }
    };

    previous = (e: React.MouseEvent) => {
        this.pageChange(e, this.state.page - 1)
    };

    next = (e: React.MouseEvent) => {
        this.pageChange(e, this.state.page + 1)
    };

    link = (page: number) => {
        return (<li key={page} style={{cursor: "pointer"}}
                    className={"pagination__item " + (this.state.page === page ? "pagination__item--current" : "")}><a
            onClick={(e) => this.pageChange(e, page)} className="pagination__link"
            aria-label={"Go to page " + page}>{page}</a></li>)
    };

    render() {
        return (
            <nav className="pagination "
                 arial-label={"Pagination (Page " + this.state.page + " of " + this.state.maxPage + ")"}>
                <div className="pagination__position" style={{marginBottom:'12px'}}>Page {this.state.page} of {this.state.maxPage}</div>
                <ul className="pagination__items" style={{marginLeft:'0px'}}>
                    {this.state.page > 1 &&
                    <li className="pagination__item pagination__item--previous" style={{backgroundColor:"#c1c9ca2e"}}>
                        <a style={{cursor: "pointer"}} onClick={(e) => this.previous(e)} className="pagination__link"
                           rel="prev" aria-label="Go to the previous page">Previous</a>
                    </li>
                    }
                    {this.link(1)}
                    {this.createLinks()}
                    {this.link(this.state.maxPage)}
                    {this.state.page < this.state.maxPage &&
                    <li className="pagination__item pagination__item--next" style={{backgroundColor:"#c1c9ca2e"}}>
                        <a style={{cursor: "pointer"}} onClick={(e) => this.next(e)} className="pagination__link"
                           rel="next" aria-label="Go to the next page">Next</a>
                    </li>
                    }
                </ul>
            </nav>
        );
    }
}
