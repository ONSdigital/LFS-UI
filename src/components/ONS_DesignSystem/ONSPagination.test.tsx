import React from "react";
import Enzyme, {mount, shallow} from "enzyme";
import {ONSPagination} from "./ONSPagination";
import Adapter from "enzyme-adapter-react-16";
import sinon from "sinon";

describe("ONS Pagination Test", () => {
    Enzyme.configure({ adapter: new Adapter() });

    const Props = {
        listLength: 10,
        count: 10,
        pageChange: sinon.spy()
    }

    const linkProps = {
        listLength: 10,
        count: 99,
        pageChange: sinon.spy()
    }

    const nextProps = { 
        listLength: 10,
        count: 30,
        pageChange: sinon.spy()
    }

    const prevProps = {
        listLength: 10,
        count: 99,
        pageChange: sinon.spy()
    }

    const line24Props = {
        listLength: 100,
        count: 10,
        pageChange: sinon.spy()
    }

    const maxProps = {
        listLength: 10,
        count: 20,
        pageChange: sinon.spy()
    }

    const emptyProps = {
        
    }



    function wrapper(render: any, props: any) {
        return render(
            <ONSPagination
                listLength={props.listLength}
                count={props.count}
                pageChange={props.pageChange}>
            </ONSPagination>
            
        )
    }

    it("matches Snapshot", () => {
        expect(wrapper(shallow, Props)).toMatchSnapshot()
    });

    it("should render correctly", () => expect(wrapper(shallow, Props).exists()).toEqual(true));

    it("should render correctly", () => expect(wrapper(shallow, line24Props).exists()).toEqual(true));

    it('simulates click events', () => {
        //next button (clicked twice for test coverage)
        wrapper(mount, nextProps).find('li.pagination__item--next').find('button').simulate('click').simulate('click');
        expect(nextProps.pageChange).toHaveProperty('callCount', 2);

        //click a link button
        let thisWrapper = wrapper(mount, linkProps)
        thisWrapper.find('li.pagination__item--next').find('button').simulate('click')
        thisWrapper.find('li.pagination__item--current').find('button').simulate('click');
        expect(linkProps.pageChange).toHaveProperty('callCount', 1);

        // prev button (next button has to be clicked for prev button to show)
        let prevWrapper = wrapper(mount, prevProps)
        prevWrapper.find('li.pagination__item--next').find('button').simulate('click')
        prevWrapper.find('li.pagination__item--previous').find('button').simulate('click')
        expect(prevProps.pageChange).toHaveProperty('callCount', 2);
    })

    it('filters the table so the amount of records and therefore the number of pages changes', () =>{
        // Checked that the prop change passes through (this test was really for the filtering)
        // Theres a maxPage condition in the component which only applies if the props change i.e. the list is filtered. 
        let thisWrapper = wrapper(mount, linkProps)
        let count = thisWrapper.find('ONSPagination').getElement().props.count
        thisWrapper.setProps({ count: 24 })
        expect(thisWrapper.find('ONSPagination').getElement().props.count === count).toBeFalsy
    })
})