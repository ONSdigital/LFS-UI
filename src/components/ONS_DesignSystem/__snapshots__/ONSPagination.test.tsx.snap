// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ONS Pagination Test matches Snapshot 1`] = `ShallowWrapper {}`;
